// catbuffer-typescript/dist/AccountStateBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountStateBuilder = void 0;
const AccountKeyFlagsDto_1 = require("./AccountKeyFlagsDto");
const AccountStateFormatDto_1 = require("./AccountStateFormatDto");
const AddressDto_1 = require("./AddressDto");
const GeneratorUtils_1 = require("./GeneratorUtils");
const HeightActivityBucketsBuilder_1 = require("./HeightActivityBucketsBuilder");
const HeightDto_1 = require("./HeightDto");
const ImportanceSnapshotBuilder_1 = require("./ImportanceSnapshotBuilder");
const KeyDto_1 = require("./KeyDto");
const MosaicBuilder_1 = require("./MosaicBuilder");
const MosaicIdDto_1 = require("./MosaicIdDto");
const VotingKeyDto_1 = require("./VotingKeyDto");
class AccountStateBuilder {
    constructor(address, addressHeight, publicKey, publicKeyHeight, accountType, supplementalAccountKeysMask, currencyMosaicId, balances, linkedPublicKey, vrfPublicKey, votingPublicKey, nodePublicKey, importanceSnapshots, activityBuckets) {
        this.address = address;
        this.addressHeight = addressHeight;
        this.publicKey = publicKey;
        this.publicKeyHeight = publicKeyHeight;
        this.accountType = accountType;
        this.supplementalAccountKeysMask = supplementalAccountKeysMask;
        this.linkedPublicKey = linkedPublicKey;
        this.vrfPublicKey = vrfPublicKey;
        this.votingPublicKey = votingPublicKey;
        this.nodePublicKey = nodePublicKey;
        this.importanceSnapshots = importanceSnapshots;
        this.activityBuckets = activityBuckets;
        this.currencyMosaicId = currencyMosaicId;
        this.balances = balances;
        if (importanceSnapshots) {
            this.format = AccountStateFormatDto_1.AccountStateFormatDto.HIGH_VALUE;
        }
        else {
            this.format = AccountStateFormatDto_1.AccountStateFormatDto.HIGH_VALUE;
        }
    }
    static loadFromBinary(payload) {
        const byteArray = Array.from(payload);
        const address = AddressDto_1.AddressDto.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, address.getSize());
        const addressHeight = HeightDto_1.HeightDto.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, addressHeight.getSize());
        const publicKey = KeyDto_1.KeyDto.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, publicKey.getSize());
        const publicKeyHeight = HeightDto_1.HeightDto.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, publicKeyHeight.getSize());
        const accountType = GeneratorUtils_1.GeneratorUtils.bufferToUint(GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), 1));
        byteArray.splice(0, 1);
        const format = GeneratorUtils_1.GeneratorUtils.bufferToUint(GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), 1));
        byteArray.splice(0, 1);
        const supplementalAccountKeysMask = GeneratorUtils_1.GeneratorUtils.bufferToUint(GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), 1));
        byteArray.splice(0, 1);
        const supplementalAccountKeysMaskConditionBytes = Uint8Array.from(byteArray.slice(0, 32));
        byteArray.splice(0, 32);
        const formatConditionBytes = Uint8Array.from(byteArray.slice(0, 1));
        byteArray.splice(0, 1);
        const currencyMosaicId = MosaicIdDto_1.MosaicIdDto.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, currencyMosaicId.getSize());
        const balancesCount = GeneratorUtils_1.GeneratorUtils.bufferToUint(GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), 2));
        byteArray.splice(0, 2);
        const balances = [];
        for (let i = 0; i < (Array.isArray(balancesCount) ? GeneratorUtils_1.GeneratorUtils.compact(balancesCount) : balancesCount); i++) {
            const item = MosaicBuilder_1.MosaicBuilder.loadFromBinary(Uint8Array.from(byteArray));
            balances.push(item);
            byteArray.splice(0, item.getSize());
        }
        let linkedPublicKey;
        if (supplementalAccountKeysMask & AccountKeyFlagsDto_1.AccountKeyFlagsDto.LINKED) {
            linkedPublicKey = KeyDto_1.KeyDto.loadFromBinary(supplementalAccountKeysMaskConditionBytes);
        }
        let vrfPublicKey;
        if (supplementalAccountKeysMask & AccountKeyFlagsDto_1.AccountKeyFlagsDto.VRF) {
            vrfPublicKey = KeyDto_1.KeyDto.loadFromBinary(supplementalAccountKeysMaskConditionBytes);
        }
        let votingPublicKey;
        if (supplementalAccountKeysMask & AccountKeyFlagsDto_1.AccountKeyFlagsDto.VOTING) {
            votingPublicKey = VotingKeyDto_1.VotingKeyDto.loadFromBinary(supplementalAccountKeysMaskConditionBytes);
        }
        let nodePublicKey;
        if (supplementalAccountKeysMask & AccountKeyFlagsDto_1.AccountKeyFlagsDto.NODE) {
            nodePublicKey = KeyDto_1.KeyDto.loadFromBinary(supplementalAccountKeysMaskConditionBytes);
        }
        let importanceSnapshots;
        if (format === AccountStateFormatDto_1.AccountStateFormatDto.HIGH_VALUE) {
            importanceSnapshots = ImportanceSnapshotBuilder_1.ImportanceSnapshotBuilder.loadFromBinary(formatConditionBytes);
        }
        let activityBuckets;
        if (format === AccountStateFormatDto_1.AccountStateFormatDto.HIGH_VALUE) {
            activityBuckets = HeightActivityBucketsBuilder_1.HeightActivityBucketsBuilder.loadFromBinary(formatConditionBytes);
        }
        return new AccountStateBuilder(address, addressHeight, publicKey, publicKeyHeight, accountType, supplementalAccountKeysMask, currencyMosaicId, balances, linkedPublicKey, vrfPublicKey, votingPublicKey, nodePublicKey, importanceSnapshots, activityBuckets);
    }
    getAddress() {
        return this.address;
    }
    getAddressHeight() {
        return this.addressHeight;
    }
    getPublicKey() {
        return this.publicKey;
    }
    getPublicKeyHeight() {
        return this.publicKeyHeight;
    }
    getAccountType() {
        return this.accountType;
    }
    getFormat() {
        return this.format;
    }
    getSupplementalAccountKeysMask() {
        return this.supplementalAccountKeysMask;
    }
    getLinkedPublicKey() {
        if (this.supplementalAccountKeysMask & AccountKeyFlagsDto_1.AccountKeyFlagsDto.LINKED) {
            throw new Error('supplementalAccountKeysMask is not set to LINKED.');
        }
        return this.linkedPublicKey;
    }
    getVrfPublicKey() {
        if (this.supplementalAccountKeysMask & AccountKeyFlagsDto_1.AccountKeyFlagsDto.VRF) {
            throw new Error('supplementalAccountKeysMask is not set to VRF.');
        }
        return this.vrfPublicKey;
    }
    getVotingPublicKey() {
        if (this.supplementalAccountKeysMask & AccountKeyFlagsDto_1.AccountKeyFlagsDto.VOTING) {
            throw new Error('supplementalAccountKeysMask is not set to VOTING.');
        }
        return this.votingPublicKey;
    }
    getNodePublicKey() {
        if (this.supplementalAccountKeysMask & AccountKeyFlagsDto_1.AccountKeyFlagsDto.NODE) {
            throw new Error('supplementalAccountKeysMask is not set to NODE.');
        }
        return this.nodePublicKey;
    }
    getImportanceSnapshots() {
        if (this.format !== AccountStateFormatDto_1.AccountStateFormatDto.HIGH_VALUE) {
            throw new Error('format is not set to HIGH_VALUE.');
        }
        return this.importanceSnapshots;
    }
    getActivityBuckets() {
        if (this.format !== AccountStateFormatDto_1.AccountStateFormatDto.HIGH_VALUE) {
            throw new Error('format is not set to HIGH_VALUE.');
        }
        return this.activityBuckets;
    }
    getCurrencyMosaicId() {
        return this.currencyMosaicId;
    }
    getBalances() {
        return this.balances;
    }
    getSize() {
        let size = 0;
        size += this.address.getSize();
        size += this.addressHeight.getSize();
        size += this.publicKey.getSize();
        size += this.publicKeyHeight.getSize();
        size += 1;
        size += 1;
        size += 1;
        if (this.supplementalAccountKeysMask & AccountKeyFlagsDto_1.AccountKeyFlagsDto.LINKED) {
            size += this.linkedPublicKey.getSize();
        }
        if (this.supplementalAccountKeysMask & AccountKeyFlagsDto_1.AccountKeyFlagsDto.VRF) {
            size += this.vrfPublicKey.getSize();
        }
        if (this.supplementalAccountKeysMask & AccountKeyFlagsDto_1.AccountKeyFlagsDto.VOTING) {
            size += this.votingPublicKey.getSize();
        }
        if (this.supplementalAccountKeysMask & AccountKeyFlagsDto_1.AccountKeyFlagsDto.NODE) {
            size += this.nodePublicKey.getSize();
        }
        if (this.format === AccountStateFormatDto_1.AccountStateFormatDto.HIGH_VALUE) {
            size += this.importanceSnapshots.getSize();
        }
        if (this.format === AccountStateFormatDto_1.AccountStateFormatDto.HIGH_VALUE) {
            size += this.activityBuckets.getSize();
        }
        size += this.currencyMosaicId.getSize();
        size += 2;
        this.balances.forEach((o) => size += o.getSize());
        return size;
    }
    serialize() {
        let newArray = Uint8Array.from([]);
        const addressBytes = this.address.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, addressBytes);
        const addressHeightBytes = this.addressHeight.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, addressHeightBytes);
        const publicKeyBytes = this.publicKey.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, publicKeyBytes);
        const publicKeyHeightBytes = this.publicKeyHeight.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, publicKeyHeightBytes);
        const accountTypeBytes = GeneratorUtils_1.GeneratorUtils.uintToBuffer(this.accountType, 1);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, accountTypeBytes);
        const formatBytes = GeneratorUtils_1.GeneratorUtils.uintToBuffer(this.format, 1);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, formatBytes);
        const supplementalAccountKeysMaskBytes = GeneratorUtils_1.GeneratorUtils.uintToBuffer(this.getSupplementalAccountKeysMask(), 1);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, supplementalAccountKeysMaskBytes);
        if (this.supplementalAccountKeysMask & AccountKeyFlagsDto_1.AccountKeyFlagsDto.LINKED) {
            const linkedPublicKeyBytes = this.linkedPublicKey.serialize();
            newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, linkedPublicKeyBytes);
        }
        if (this.supplementalAccountKeysMask & AccountKeyFlagsDto_1.AccountKeyFlagsDto.VRF) {
            const vrfPublicKeyBytes = this.vrfPublicKey.serialize();
            newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, vrfPublicKeyBytes);
        }
        if (this.supplementalAccountKeysMask & AccountKeyFlagsDto_1.AccountKeyFlagsDto.VOTING) {
            const votingPublicKeyBytes = this.votingPublicKey.serialize();
            newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, votingPublicKeyBytes);
        }
        if (this.supplementalAccountKeysMask & AccountKeyFlagsDto_1.AccountKeyFlagsDto.NODE) {
            const nodePublicKeyBytes = this.nodePublicKey.serialize();
            newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, nodePublicKeyBytes);
        }
        if (this.format === AccountStateFormatDto_1.AccountStateFormatDto.HIGH_VALUE) {
            const importanceSnapshotsBytes = this.importanceSnapshots.serialize();
            newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, importanceSnapshotsBytes);
        }
        if (this.format === AccountStateFormatDto_1.AccountStateFormatDto.HIGH_VALUE) {
            const activityBucketsBytes = this.activityBuckets.serialize();
            newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, activityBucketsBytes);
        }
        const currencyMosaicIdBytes = this.currencyMosaicId.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, currencyMosaicIdBytes);
        const balancesCountBytes = GeneratorUtils_1.GeneratorUtils.uintToBuffer(this.balances.length, 2);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, balancesCountBytes);
        this.balances.forEach((item) => {
            const balancesBytes = item.serialize();
            newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, balancesBytes);
        });
        return newArray;
    }
}
exports.AccountStateBuilder = AccountStateBuilder;
