// symbol-openapi-typescript-node-client/model/embeddedTransactionInfoDTO.ts
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.0.20.34
 *
 * The version of the OpenAPI document: 0.8.11
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { RequestFile } from '../api';
import { EmbeddedAccountAddressRestrictionTransactionDTO } from './embeddedAccountAddressRestrictionTransactionDTO';
import { EmbeddedAccountKeyLinkTransactionDTO } from './embeddedAccountKeyLinkTransactionDTO';
import { EmbeddedAccountMetadataTransactionDTO } from './embeddedAccountMetadataTransactionDTO';
import { EmbeddedAccountMosaicRestrictionTransactionDTO } from './embeddedAccountMosaicRestrictionTransactionDTO';
import { EmbeddedAccountOperationRestrictionTransactionDTO } from './embeddedAccountOperationRestrictionTransactionDTO';
import { EmbeddedAddressAliasTransactionDTO } from './embeddedAddressAliasTransactionDTO';
import { EmbeddedHashLockTransactionDTO } from './embeddedHashLockTransactionDTO';
import { EmbeddedMosaicAddressRestrictionTransactionDTO } from './embeddedMosaicAddressRestrictionTransactionDTO';
import { EmbeddedMosaicAliasTransactionDTO } from './embeddedMosaicAliasTransactionDTO';
import { EmbeddedMosaicDefinitionTransactionDTO } from './embeddedMosaicDefinitionTransactionDTO';
import { EmbeddedMosaicGlobalRestrictionTransactionDTO } from './embeddedMosaicGlobalRestrictionTransactionDTO';
import { EmbeddedMosaicMetadataTransactionDTO } from './embeddedMosaicMetadataTransactionDTO';
import { EmbeddedMosaicSupplyChangeTransactionDTO } from './embeddedMosaicSupplyChangeTransactionDTO';
import { EmbeddedMultisigAccountModificationTransactionDTO } from './embeddedMultisigAccountModificationTransactionDTO';
import { EmbeddedNamespaceMetadataTransactionDTO } from './embeddedNamespaceMetadataTransactionDTO';
import { EmbeddedNamespaceRegistrationTransactionDTO } from './embeddedNamespaceRegistrationTransactionDTO';
import { EmbeddedNodeKeyLinkTransactionDTO } from './embeddedNodeKeyLinkTransactionDTO';
import { EmbeddedSecretLockTransactionDTO } from './embeddedSecretLockTransactionDTO';
import { EmbeddedSecretProofTransactionDTO } from './embeddedSecretProofTransactionDTO';
import { EmbeddedTransactionMetaDTO } from './embeddedTransactionMetaDTO';
import { EmbeddedTransferTransactionDTO } from './embeddedTransferTransactionDTO';
import { EmbeddedVotingKeyLinkTransactionDTO } from './embeddedVotingKeyLinkTransactionDTO';
import { EmbeddedVrfKeyLinkTransactionDTO } from './embeddedVrfKeyLinkTransactionDTO';

export class EmbeddedTransactionInfoDTO {
    'meta': EmbeddedTransactionMetaDTO;
    'transaction': EmbeddedAccountKeyLinkTransactionDTO | EmbeddedNodeKeyLinkTransactionDTO | EmbeddedVrfKeyLinkTransactionDTO | EmbeddedVotingKeyLinkTransactionDTO | EmbeddedHashLockTransactionDTO | EmbeddedSecretLockTransactionDTO | EmbeddedSecretProofTransactionDTO | EmbeddedAccountMetadataTransactionDTO | EmbeddedMosaicMetadataTransactionDTO | EmbeddedNamespaceMetadataTransactionDTO | EmbeddedMosaicDefinitionTransactionDTO | EmbeddedMosaicSupplyChangeTransactionDTO | EmbeddedNamespaceRegistrationTransactionDTO | EmbeddedAddressAliasTransactionDTO | EmbeddedMosaicAliasTransactionDTO | EmbeddedMultisigAccountModificationTransactionDTO | EmbeddedAccountAddressRestrictionTransactionDTO | EmbeddedAccountMosaicRestrictionTransactionDTO | EmbeddedAccountOperationRestrictionTransactionDTO | EmbeddedMosaicGlobalRestrictionTransactionDTO | EmbeddedMosaicAddressRestrictionTransactionDTO | EmbeddedTransferTransactionDTO;

    static discriminator: string | undefined = undefined;

    static attributeTypeMap: Array<{name: string, baseName: string, type: string}> = [
        {
            "name": "meta",
            "baseName": "meta",
            "type": "EmbeddedTransactionMetaDTO"
        },
        {
            "name": "transaction",
            "baseName": "transaction",
            "type": "EmbeddedAccountKeyLinkTransactionDTO | EmbeddedNodeKeyLinkTransactionDTO | EmbeddedVrfKeyLinkTransactionDTO | EmbeddedVotingKeyLinkTransactionDTO | EmbeddedHashLockTransactionDTO | EmbeddedSecretLockTransactionDTO | EmbeddedSecretProofTransactionDTO | EmbeddedAccountMetadataTransactionDTO | EmbeddedMosaicMetadataTransactionDTO | EmbeddedNamespaceMetadataTransactionDTO | EmbeddedMosaicDefinitionTransactionDTO | EmbeddedMosaicSupplyChangeTransactionDTO | EmbeddedNamespaceRegistrationTransactionDTO | EmbeddedAddressAliasTransactionDTO | EmbeddedMosaicAliasTransactionDTO | EmbeddedMultisigAccountModificationTransactionDTO | EmbeddedAccountAddressRestrictionTransactionDTO | EmbeddedAccountMosaicRestrictionTransactionDTO | EmbeddedAccountOperationRestrictionTransactionDTO | EmbeddedMosaicGlobalRestrictionTransactionDTO | EmbeddedMosaicAddressRestrictionTransactionDTO | EmbeddedTransferTransactionDTO"
        }    ];

    static getAttributeTypeMap() {
        return EmbeddedTransactionInfoDTO.attributeTypeMap;
    }
}
