// symbol-openapi-typescript-node-client/model/transactionInfoDTO.ts
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.0.20.34
 *
 * The version of the OpenAPI document: 0.8.11
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { RequestFile } from '../api';
import { AccountAddressRestrictionTransactionDTO } from './accountAddressRestrictionTransactionDTO';
import { AccountKeyLinkTransactionDTO } from './accountKeyLinkTransactionDTO';
import { AccountMetadataTransactionDTO } from './accountMetadataTransactionDTO';
import { AccountMosaicRestrictionTransactionDTO } from './accountMosaicRestrictionTransactionDTO';
import { AccountOperationRestrictionTransactionDTO } from './accountOperationRestrictionTransactionDTO';
import { AddressAliasTransactionDTO } from './addressAliasTransactionDTO';
import { AggregateTransactionDTO } from './aggregateTransactionDTO';
import { HashLockTransactionDTO } from './hashLockTransactionDTO';
import { MosaicAddressRestrictionTransactionDTO } from './mosaicAddressRestrictionTransactionDTO';
import { MosaicAliasTransactionDTO } from './mosaicAliasTransactionDTO';
import { MosaicDefinitionTransactionDTO } from './mosaicDefinitionTransactionDTO';
import { MosaicGlobalRestrictionTransactionDTO } from './mosaicGlobalRestrictionTransactionDTO';
import { MosaicMetadataTransactionDTO } from './mosaicMetadataTransactionDTO';
import { MosaicSupplyChangeTransactionDTO } from './mosaicSupplyChangeTransactionDTO';
import { MultisigAccountModificationTransactionDTO } from './multisigAccountModificationTransactionDTO';
import { NamespaceMetadataTransactionDTO } from './namespaceMetadataTransactionDTO';
import { NamespaceRegistrationTransactionDTO } from './namespaceRegistrationTransactionDTO';
import { NodeKeyLinkTransactionDTO } from './nodeKeyLinkTransactionDTO';
import { SecretLockTransactionDTO } from './secretLockTransactionDTO';
import { SecretProofTransactionDTO } from './secretProofTransactionDTO';
import { TransactionMetaDTO } from './transactionMetaDTO';
import { TransferTransactionDTO } from './transferTransactionDTO';
import { VotingKeyLinkTransactionDTO } from './votingKeyLinkTransactionDTO';
import { VrfKeyLinkTransactionDTO } from './vrfKeyLinkTransactionDTO';

export class TransactionInfoDTO {
    'meta': TransactionMetaDTO;
    'transaction': AccountKeyLinkTransactionDTO | NodeKeyLinkTransactionDTO | VrfKeyLinkTransactionDTO | VotingKeyLinkTransactionDTO | AggregateTransactionDTO | HashLockTransactionDTO | SecretLockTransactionDTO | SecretProofTransactionDTO | AccountMetadataTransactionDTO | MosaicMetadataTransactionDTO | NamespaceMetadataTransactionDTO | MosaicDefinitionTransactionDTO | MosaicSupplyChangeTransactionDTO | NamespaceRegistrationTransactionDTO | AddressAliasTransactionDTO | MosaicAliasTransactionDTO | MultisigAccountModificationTransactionDTO | AccountAddressRestrictionTransactionDTO | AccountMosaicRestrictionTransactionDTO | AccountOperationRestrictionTransactionDTO | MosaicGlobalRestrictionTransactionDTO | MosaicAddressRestrictionTransactionDTO | TransferTransactionDTO;

    static discriminator: string | undefined = undefined;

    static attributeTypeMap: Array<{name: string, baseName: string, type: string}> = [
        {
            "name": "meta",
            "baseName": "meta",
            "type": "TransactionMetaDTO"
        },
        {
            "name": "transaction",
            "baseName": "transaction",
            "type": "AccountKeyLinkTransactionDTO | NodeKeyLinkTransactionDTO | VrfKeyLinkTransactionDTO | VotingKeyLinkTransactionDTO | AggregateTransactionDTO | HashLockTransactionDTO | SecretLockTransactionDTO | SecretProofTransactionDTO | AccountMetadataTransactionDTO | MosaicMetadataTransactionDTO | NamespaceMetadataTransactionDTO | MosaicDefinitionTransactionDTO | MosaicSupplyChangeTransactionDTO | NamespaceRegistrationTransactionDTO | AddressAliasTransactionDTO | MosaicAliasTransactionDTO | MultisigAccountModificationTransactionDTO | AccountAddressRestrictionTransactionDTO | AccountMosaicRestrictionTransactionDTO | AccountOperationRestrictionTransactionDTO | MosaicGlobalRestrictionTransactionDTO | MosaicAddressRestrictionTransactionDTO | TransferTransactionDTO"
        }    ];

    static getAttributeTypeMap() {
        return TransactionInfoDTO.attributeTypeMap;
    }
}
